# <center> Unlocking Business Potential through Transactional Data </center>


##  Business Understanding


In this project, our goal is to transform untapped transactional data from the year 2019 into actionable insights that drive sales and operational efficiency for our client. By leveraging advanced business intelligence solutions, we aim to uncover opportunities and trends within the data, providing strategic recommendations to optimize business processes and maximize revenue.

###  Business Objective
The primary objective of this business intelligence initiative is to harness the potential of the client's 2019 transactional data to optimize operations, drive targeted sales strategies, and ultimately maximize revenue. 

###  Hypothesis

**Null Hypothesis (H0):** There is no significant relationship between the type of product and the quantity ordered.
 
**Alternative Hypothesis (H1):** There is a significant relationship between the type of product and the quantity ordered. 


### Questions
1. How much money did we make this year? 

2. Can we identify any seasonality in the  sales? 

3. What are our best and worst-selling products? 

4. How do sales compare to previous months or weeks? 

5. Which cities are our products delivered to most? 

6. How do product categories compare in revenue generated and quantities  ordered? 

### Install required packages

In [2]:
#Libraries for sql
import pyodbc 
from dotenv import dotenv_values #import the dotenv_values function from the dotenv package
import warnings 
warnings.filterwarnings('ignore')

#libraries for handling data
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)

#libraries for visulation
import seaborn as sns
from matplotlib import pyplot as plt
import plotly.offline as offline
from plotly.offline import init_notebook_mode, iplot
import plotly.express as px
import plotly.graph_objects as go
import plotly.subplots as sp
offline.init_notebook_mode(connected=True) # Configure Plotly to run 

### Create a connection by accessing connection string with defined environment variables


In [3]:
# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values('.env')

In [4]:
# Get the values for the credentials you set in the '.env' file
database = environment_variables.get("DATABASE")
server = environment_variables.get("SERVER")
username = environment_variables.get("USERNAME")
password = environment_variables.get("PASSWORD")

connection_string = f"DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}"

In [6]:
# Use the connect method of the pyodbc library and pass in the connection string.
connection = pyodbc.connect(connection_string)

OperationalError: ('08001', '[08001] [Microsoft][ODBC SQL Server Driver][DBNETLIB]SQL Server does not exist or access denied. (17) (SQLDriverConnect); [08001] [Microsoft][ODBC SQL Server Driver][DBNETLIB]ConnectionOpen (Connect()). (53)')

In [ ]:
# Now the sql query to get the data is what what you see below. 
query_jul = "Select * from dbo.Sales_July_2019"
query_aug = "Select * from dbo.Sales_August_2019"
query_sep = "Select * from dbo.Sales_September_2019"
query_oct = "Select * from dbo.Sales_October_2019"
query_nov = "Select * from dbo.Sales_November_2019"
query_dec = "Select * from dbo.Sales_December_2019"


## Data Understanding


###  Resources availability 
Data for this project

In this project, the dataset resides in 2 places. For ease of access  the datasetshave been made available in my GitHub Repository 

- The first 6 months of the dataset was extracted from one drive
- The next 6 months was extracted from a database


###  Column names and description

**Order ID**: number to represent customer ID

**Product**:Type of Product

**Quantity_Ordered**: Number of products ordered

**Price_Each**: Price of each product

**Order_Date**: Date product was ordered

**Purchase Address**: Address product was purchased


### Data Loading

We load our datasets

In [ ]:
##load datasets in csv
df_jan=pd.read_csv(r"Datasets\Sales_January_2019.csv")
df_feb=pd.read_csv(r"Datasets\Sales_February_2019.csv")
df_mar=pd.read_csv(r"Datasets\Sales_March_2019.csv")
df_apr=pd.read_csv(r"Datasets\Sales_April_2019.csv")
df_may=pd.read_csv(r"Datasets\Sales_May_2019.csv")
df_jun=pd.read_csv(r"Datasets\Sales_June_2019.csv")

#load datasets in database
df_jul=pd.read_sql(query_jul,connection)
df_aug=pd.read_sql(query_aug,connection)
df_sep=pd.read_sql(query_sep,connection)
df_oct=pd.read_sql(query_oct,connection)
df_nov=pd.read_sql(query_nov,connection)
df_dec=pd.read_sql(query_dec,connection)

In [ ]:
#we will put convert the dataset in the database to csv
df_jul.to_csv("July Sales.csv",index=False)
df_aug.to_csv("August Sales.csv",index=False)
df_sep.to_csv("September Sales.csv",index=False)
df_oct.to_csv("october Sales.csv",index=False)
df_nov.to_csv("November Sales.csv",index=False)
df_dec.to_csv("December Sales.csv",index=False)

## View datasets

In [ ]:
#view january dataset
df_jan.head()

Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [ ]:
#view february dataset
df_feb.head()

Order ID                   Product Quantity Ordered Price Each  \
0   150502                    iPhone                1        700   
1   150503     AA Batteries (4-pack)                1       3.84   
2   150504    27in 4K Gaming Monitor                1     389.99   
3   150505  Lightning Charging Cable                1      14.95   
4   150506     AA Batteries (4-pack)                2       3.84   

       Order Date                     Purchase Address  
0  02/18/19 01:35    866 Spruce St, Portland, ME 04101  
1  02/13/19 07:24  18 13th St, San Francisco, CA 94016  
2  02/18/19 09:46   52 6th St, New York City, NY 10001  
3  02/02/19 16:47     129 Cherry St, Atlanta, GA 30301  
4  02/28/19 20:32    548 Lincoln St, Seattle, WA 98101

In [ ]:
#view march
df_mar.head()

Order ID                     Product Quantity Ordered Price Each  \
0   162009                      iPhone                1        700   
1   162009    Lightning Charging Cable                1      14.95   
2   162009            Wired Headphones                2      11.99   
3   162010  Bose SoundSport Headphones                1      99.99   
4   162011      34in Ultrawide Monitor                1     379.99   

       Order Date                      Purchase Address  
0  03/28/19 20:59       942 Church St, Austin, TX 73301  
1  03/28/19 20:59       942 Church St, Austin, TX 73301  
2  03/28/19 20:59       942 Church St, Austin, TX 73301  
3  03/17/19 05:39  261 10th St, San Francisco, CA 94016  
4  03/10/19 00:01  764 13th St, San Francisco, CA 94016

In [ ]:
#view april
df_apr.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   

       Order Date                      Purchase Address  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001  
1             NaN                                   NaN  
2  04/07/19 22:30     682 Chestnut St, Boston, MA 02215  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  
4  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001

In [ ]:
#view may
df_may.head()

Order ID                 Product Quantity Ordered Price Each  \
0   194095        Wired Headphones                1      11.99   
1   194096   AA Batteries (4-pack)                1       3.84   
2   194097        27in FHD Monitor                1     149.99   
3   194098        Wired Headphones                1      11.99   
4   194099  AAA Batteries (4-pack)                2       2.99   

       Order Date                         Purchase Address  
0  05/16/19 17:14      669 2nd St, New York City, NY 10001  
1  05/19/19 14:43          844 Walnut St, Dallas, TX 75001  
2  05/24/19 11:36  164 Madison St, New York City, NY 10001  
3  05/02/19 20:40          622 Meadow St, Dallas, TX 75001  
4  05/11/19 22:55          17 Church St, Seattle, WA 98101

In [ ]:
#view june
df_jun.head()

Order ID                     Product Quantity Ordered Price Each  \
0   209921        USB-C Charging Cable                1      11.95   
1   209922          Macbook Pro Laptop                1     1700.0   
2   209923             ThinkPad Laptop                1     999.99   
3   209924            27in FHD Monitor                1     149.99   
4   209925  Bose SoundSport Headphones                1      99.99   

       Order Date                       Purchase Address  
0  06/23/19 19:34      950 Walnut St, Portland, ME 04101  
1  06/30/19 10:05     80 4th St, San Francisco, CA 94016  
2  06/24/19 20:18  402 Jackson St, Los Angeles, CA 90001  
3  06/05/19 10:21         560 10th St, Seattle, WA 98101  
4  06/25/19 18:58    545 2nd St, San Francisco, CA 94016

In [ ]:
#view july
df_jul.head()

Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

                    Order_Date                  Purchase_Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

In [ ]:
#view august
df_aug.head()

Order_ID                     Product  Quantity_Ordered  Price_Each  \
0  236670.0            Wired Headphones               2.0   11.990000   
1  236671.0  Bose SoundSport Headphones               1.0   99.989998   
2  236672.0                      iPhone               1.0  700.000000   
3  236673.0       AA Batteries (4-pack)               2.0    3.840000   
4  236674.0       AA Batteries (4-pack)               2.0    3.840000   

                    Order_Date                      Purchase_Address  
0  2031-08-19 22:21:00.0000000      359 Spruce St, Seattle, WA 98101  
1  2015-08-19 15:11:00.0000000        492 Ridge St, Dallas, TX 75001  
2  2006-08-19 14:40:00.0000000        149 7th St, Portland, OR 97035  
3  2029-08-19 20:59:00.0000000     631 2nd St, Los Angeles, CA 90001  
4  2015-08-19 19:53:00.0000000  736 14th St, New York City, NY 10001

In [ ]:
#view september
df_sep.head()

Order_ID                Product  Quantity_Ordered  Price_Each  \
0  248151.0  AA Batteries (4-pack)               4.0    3.840000   
1  248152.0   USB-C Charging Cable               2.0   11.950000   
2  248153.0   USB-C Charging Cable               1.0   11.950000   
3  248154.0       27in FHD Monitor               1.0  149.990005   
4  248155.0   USB-C Charging Cable               1.0   11.950000   

                    Order_Date                       Purchase_Address  
0  2017-09-19 14:44:00.0000000    380 North St, Los Angeles, CA 90001  
1  2029-09-19 10:19:00.0000000           511 8th St, Austin, TX 73301  
2  2016-09-19 17:48:00.0000000  151 Johnson St, Los Angeles, CA 90001  
3  2027-09-19 07:52:00.0000000      355 Hickory St, Seattle, WA 98101  
4  2001-09-19 19:03:00.0000000          125 5th St, Atlanta, GA 30301

In [ ]:
#view october
df_oct.head()


Order_ID                 Product  Quantity_Ordered  Price_Each  \
0  259358.0  34in Ultrawide Monitor               1.0  379.989990   
1  259359.0  27in 4K Gaming Monitor               1.0  389.989990   
2  259360.0  AAA Batteries (4-pack)               2.0    2.990000   
3  259361.0        27in FHD Monitor               1.0  149.990005   
4  259362.0        Wired Headphones               1.0   11.990000   

                    Order_Date                           Purchase_Address  
0  2028-10-19 10:56:00.0000000            609 Cherry St, Dallas, TX 75001  
1  2028-10-19 17:26:00.0000000          225 5th St, Los Angeles, CA 90001  
2  2024-10-19 17:20:00.0000000       967 12th St, New York City, NY 10001  
3  2014-10-19 22:26:00.0000000  628 Jefferson St, New York City, NY 10001  
4  2007-10-19 16:10:00.0000000         534 14th St, Los Angeles, CA 90001

In [ ]:
#view november
df_nov.head()


Order_ID                     Product  Quantity_Ordered  Price_Each  \
0  278797.0            Wired Headphones               1.0   11.990000   
1  278798.0        USB-C Charging Cable               2.0   11.950000   
2  278799.0    Apple Airpods Headphones               1.0  150.000000   
3  278800.0            27in FHD Monitor               1.0  149.990005   
4  278801.0  Bose SoundSport Headphones               1.0   99.989998   

                    Order_Date                      Purchase_Address  
0  2021-11-19 09:54:00.0000000   46 Park St, New York City, NY 10001  
1  2017-11-19 10:03:00.0000000      962 Hickory St, Austin, TX 73301  
2  2019-11-19 14:56:00.0000000  464 Cherry St, Los Angeles, CA 90001  
3  2025-11-19 22:24:00.0000000        649 10th St, Seattle, WA 98101  
4  2009-11-19 13:56:00.0000000         522 Hill St, Boston, MA 02215

In [ ]:
#view december
df_dec.head()

Order_ID               Product  Quantity_Ordered   Price_Each  \
0  295665.0    Macbook Pro Laptop               1.0  1700.000000   
1  295666.0    LG Washing Machine               1.0   600.000000   
2  295667.0  USB-C Charging Cable               1.0    11.950000   
3  295668.0      27in FHD Monitor               1.0   149.990005   
4  295669.0  USB-C Charging Cable               1.0    11.950000   

                    Order_Date                        Purchase_Address  
0  2030-12-19 00:01:00.0000000  136 Church St, New York City, NY 10001  
1  2029-12-19 07:03:00.0000000     562 2nd St, New York City, NY 10001  
2  2012-12-19 18:21:00.0000000    277 Main St, New York City, NY 10001  
3  2022-12-19 15:13:00.0000000     410 6th St, San Francisco, CA 94016  
4  2018-12-19 12:38:00.0000000           43 Hill St, Atlanta, GA 30301

## Exploratory Data Analysis
Here we will analyze and investigate data sets and summarize their characteristics by using data visualization,


## Data Overview

#### Data info
Here we want to check the information the dataset contains

In [ ]:
#january dataset info
df_jan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9723 entries, 0 to 9722
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          9697 non-null   object
 1   Product           9697 non-null   object
 2   Quantity Ordered  9697 non-null   object
 3   Price Each        9697 non-null   object
 4   Order Date        9697 non-null   object
 5   Purchase Address  9697 non-null   object
dtypes: object(6)
memory usage: 455.9+ KB


## Observation on january dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 9723 enteries but each column has 9697 indicating presence of missing values

In [ ]:
#february info
df_feb.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12036 entries, 0 to 12035
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          12004 non-null  object
 1   Product           12004 non-null  object
 2   Quantity Ordered  12004 non-null  object
 3   Price Each        12004 non-null  object
 4   Order Date        12004 non-null  object
 5   Purchase Address  12004 non-null  object
dtypes: object(6)
memory usage: 564.3+ KB


## Observation on February dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 12036 enteries but each column has 12004 indicating presence of missing values

In [ ]:
#view info for march dataset
df_mar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15226 entries, 0 to 15225
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          15189 non-null  object
 1   Product           15189 non-null  object
 2   Quantity Ordered  15189 non-null  object
 3   Price Each        15189 non-null  object
 4   Order Date        15189 non-null  object
 5   Purchase Address  15189 non-null  object
dtypes: object(6)
memory usage: 713.8+ KB


## Observation on March dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 15226 enteries but each column has 15189 indicating presence of missing values

In [ ]:
#info for april
df_apr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18383 entries, 0 to 18382
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          18324 non-null  object
 1   Product           18324 non-null  object
 2   Quantity Ordered  18324 non-null  object
 3   Price Each        18324 non-null  object
 4   Order Date        18324 non-null  object
 5   Purchase Address  18324 non-null  object
dtypes: object(6)
memory usage: 861.8+ KB


## Observation on April dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 18383 enteries but each column has 18324 indicating presence of missing values

In [ ]:
#view may info
df_may.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16635 entries, 0 to 16634
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          16587 non-null  object
 1   Product           16587 non-null  object
 2   Quantity Ordered  16587 non-null  object
 3   Price Each        16587 non-null  object
 4   Order Date        16587 non-null  object
 5   Purchase Address  16587 non-null  object
dtypes: object(6)
memory usage: 779.9+ KB


## Observation on May dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 16635 enteries but each column has 16587 indicating presence of missing values

In [ ]:
#view june info
df_jun.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13622 entries, 0 to 13621
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          13579 non-null  object
 1   Product           13579 non-null  object
 2   Quantity Ordered  13579 non-null  object
 3   Price Each        13579 non-null  object
 4   Order Date        13579 non-null  object
 5   Purchase Address  13579 non-null  object
dtypes: object(6)
memory usage: 638.7+ KB


## Observation on June dataset
- Order id,Quantity_ordered,order date and price each have wrong datatypes
- There are 13622 enteries but each column has 13579 indicating presence of missing values

In [ ]:
#view info of july dataset
df_jul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14371 entries, 0 to 14370
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          14291 non-null  float64
 1   Product           14326 non-null  object 
 2   Quantity_Ordered  14291 non-null  float64
 3   Price_Each        14291 non-null  float64
 4   Order_Date        14291 non-null  object 
 5   Purchase_Address  14326 non-null  object 
dtypes: float64(3), object(3)
memory usage: 673.8+ KB


## Observation on july dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 14371 enteries but each column have enteries not up to the total indicating presence of missing values

In [ ]:
#view info for august dataset
df_aug.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12011 entries, 0 to 12010
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          11957 non-null  float64
 1   Product           11983 non-null  object 
 2   Quantity_Ordered  11957 non-null  float64
 3   Price_Each        11957 non-null  float64
 4   Order_Date        11957 non-null  object 
 5   Purchase_Address  11983 non-null  object 
dtypes: float64(3), object(3)
memory usage: 563.1+ KB


## Observation on august dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 12011 enteries but each column have enteries not up to the total indicating presence of missing values

In [ ]:
#view september info
df_sep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11686 entries, 0 to 11685
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          11629 non-null  float64
 1   Product           11646 non-null  object 
 2   Quantity_Ordered  11629 non-null  float64
 3   Price_Each        11629 non-null  float64
 4   Order_Date        11629 non-null  object 
 5   Purchase_Address  11646 non-null  object 
dtypes: float64(3), object(3)
memory usage: 547.9+ KB


## Observation on September dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 11686 enteries but each column have enteries not up to the total indicating presence of missing values

In [ ]:
#view october info
df_oct.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20379 entries, 0 to 20378
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          20284 non-null  float64
 1   Product           20317 non-null  object 
 2   Quantity_Ordered  20284 non-null  float64
 3   Price_Each        20284 non-null  float64
 4   Order_Date        20284 non-null  object 
 5   Purchase_Address  20317 non-null  object 
dtypes: float64(3), object(3)
memory usage: 955.4+ KB


## Observation on october dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 20379 enteries but each column have enteries not up to the total indicating presence of missing values

In [ ]:
#view november info
df_nov.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17661 entries, 0 to 17660
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          17580 non-null  float64
 1   Product           17616 non-null  object 
 2   Quantity_Ordered  17580 non-null  float64
 3   Price_Each        17580 non-null  float64
 4   Order_Date        17580 non-null  object 
 5   Purchase_Address  17616 non-null  object 
dtypes: float64(3), object(3)
memory usage: 828.0+ KB


## Observation on November dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 17661 enteries but each column have enteries not up to the total indicating presence of missing values

In [ ]:
#view december info
df_dec.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25117 entries, 0 to 25116
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Order_ID          24989 non-null  float64
 1   Product           25037 non-null  object 
 2   Quantity_Ordered  24989 non-null  float64
 3   Price_Each        24989 non-null  float64
 4   Order_Date        24989 non-null  object 
 5   Purchase_Address  25037 non-null  object 
dtypes: float64(3), object(3)
memory usage: 1.1+ MB


## Observation on December dataset
- Quantity_ordered,order date,order each have wrong datatypes.With the quantity ordered,it is suppose to be a numeric value but after viewing the data,a customer can't order a quantity in decimal hence we will chnage the datatype.Same with order id it has to be numeric but not in float format

- There are 25117 enteries but each column have enteries not up to the total indicating presence of missing values

## General Observation after checking info of the datasets
- In all the 12 datasets representing the 12 months,they all have wrong datatypes and missing values present
- The first 6 months,have the same datatypes across them.The next six months also have the same datatypes.
- Also the column representing purachase address,have their cities attached an we want just the cities from the data so we will discard the rest
- Some columns in the last 6 months had underscore in their names others didn't so we will rename them

## Check for missing values

In [ ]:
#view missing values in january
df_jan.isnull().sum()


Order ID            26
Product             26
Quantity Ordered    26
Price Each          26
Order Date          26
Purchase Address    26
dtype: int64

In [ ]:
df_feb.isnull().sum()

Order ID            32
Product             32
Quantity Ordered    32
Price Each          32
Order Date          32
Purchase Address    32
dtype: int64

In [ ]:
df_mar.isnull().sum()

Order ID            37
Product             37
Quantity Ordered    37
Price Each          37
Order Date          37
Purchase Address    37
dtype: int64

In [ ]:
df_apr.isnull().sum()

Order ID            59
Product             59
Quantity Ordered    59
Price Each          59
Order Date          59
Purchase Address    59
dtype: int64

In [ ]:
df_may.isnull().sum()

Order ID            48
Product             48
Quantity Ordered    48
Price Each          48
Order Date          48
Purchase Address    48
dtype: int64

In [ ]:
df_jun.isnull().sum()

Order ID            43
Product             43
Quantity Ordered    43
Price Each          43
Order Date          43
Purchase Address    43
dtype: int64

In [ ]:
df_jul.isnull().sum()

Order_ID            80
Product             45
Quantity_Ordered    80
Price_Each          80
Order_Date          80
Purchase_Address    45
dtype: int64

In [ ]:
df_aug.isnull().sum()

Order_ID            54
Product             28
Quantity_Ordered    54
Price_Each          54
Order_Date          54
Purchase_Address    28
dtype: int64

In [ ]:
df_sep.isnull().sum()

Order_ID            57
Product             40
Quantity_Ordered    57
Price_Each          57
Order_Date          57
Purchase_Address    40
dtype: int64

In [ ]:
df_oct.isnull().sum()

Order_ID            95
Product             62
Quantity_Ordered    95
Price_Each          95
Order_Date          95
Purchase_Address    62
dtype: int64

In [ ]:
df_nov.isnull().sum()

Order_ID            81
Product             45
Quantity_Ordered    81
Price_Each          81
Order_Date          81
Purchase_Address    45
dtype: int64

In [ ]:
df_dec.isnull().sum()

Order_ID            128
Product              80
Quantity_Ordered    128
Price_Each          128
Order_Date          128
Purchase_Address     80
dtype: int64

## Check for duplicates

In [ ]:
# # Assuming df_jan, df_feb, ..., df_dec are your DataFrames
# dataframes = [df_jan, df_feb, df_mar, df_apr, df_may, df_jun, df_jul, df_aug, df_sep, df_oct, df_nov, df_dec]

# # Check for missing values in each DataFrame
# for idx, df in enumerate(dataframes, start=1):
#     print(f"DataFrame {idx}:")
#     print(df.duplicated().sum())
#     print("\n" + "=" * 50 + "\n")  # Separating each DataFrame's output


DataFrame 1:
Order ID            26
Product             26
Quantity Ordered    26
Price Each          26
Order Date          26
Purchase Address    26
dtype: int64


DataFrame 2:
Order ID            32
Product             32
Quantity Ordered    32
Price Each          32
Order Date          32
Purchase Address    32
dtype: int64


DataFrame 3:
Order ID            37
Product             37
Quantity Ordered    37
Price Each          37
Order Date          37
Purchase Address    37
dtype: int64


DataFrame 4:
Order ID            59
Product             59
Quantity Ordered    59
Price Each          59
Order Date          59
Purchase Address    59
dtype: int64


DataFrame 5:
Order ID            48
Product             48
Quantity Ordered    48
Price Each          48
Order Date          48
Purchase Address    48
dtype: int64


DataFrame 6:
Order ID            43
Product             43
Quantity Ordered    43
Price Each          43
Order Date          43
Purchase Address    43
dtype: int64


Data

In [ ]:
# Assuming df_jan, df_feb, ..., df_dec are your DataFrames
dataframes = [df_jan, df_feb, df_mar, df_apr, df_may, df_jun, df_jul, df_aug, df_sep, df_oct, df_nov, df_dec]

# Check for duplicated values in each DataFrame
for month, df in zip(['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'], dataframes):
    print(f"DataFrame {month}:")
    print(df.duplicated().sum())
    print("\n" + "=" * 50 + "\n")  # Separating each DataFrame's output


### Renaming columns

## Handling Wrong datatypes